echo-telemetry/src/main/java/com/netflix/spinnaker/echo/telemetry/TelemetryEventListener.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2019 Armory, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.echo.telemetry;
+
+import com.google.protobuf.util.JsonFormat;
+import com.netflix.spinnaker.echo.events.EchoEventListener;
+import com.netflix.spinnaker.echo.model.Event;
+import com.netflix.spinnaker.proto.stats.Application;
+import com.netflix.spinnaker.proto.stats.Execution;
+import com.netflix.spinnaker.proto.stats.SpinnakerInstance;
+import com.netflix.spinnaker.proto.stats.Stage;
+import com.netflix.spinnaker.proto.stats.Status;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.stereotype.Component;
+
+@Slf4j
+@Component
+@ConditionalOnProperty("telemetry.enabled")
+public class TelemetryEventListener implements EchoEventListener {
+  private final TelemetryService telemetryService;
+
+  @Autowired
+  public TelemetryEventListener(TelemetryService telemetryService) {
+    this.telemetryService = telemetryService;
+  }
+
+  @Value("${telemetry.instanceId}")
+  String instanceId;
+
+  @Override
+  public void processEvent(Event event) {
+    try {
+      if (event.getDetails() == null
+          || !event.getDetails().getType().equals("orca:pipeline:complete")) {
+        return;
+      }
+
+      Map execution = (Map) event.content.get("execution");
+      Execution.Builder executionBuilder = getExecutionBuilder(execution);
+
+      List<Map> stages = (ArrayList<Map>) execution.get("stages");
+      for (Map stage : stages) {
+        executionBuilder.addStages(getStageBuilder(stage));
+      }
+
+      Application.Builder applicationBuilder =
+          Application.newBuilder()
+              .setId(event.details.getApplication())
+              .setExecution(executionBuilder);
+
+      SpinnakerInstance.Builder spinnakerInstance =
+          SpinnakerInstance.newBuilder().setId(instanceId).setApplication(applicationBuilder);
+
+      com.netflix.spinnaker.proto.stats.Event.Builder eventProto =
+          com.netflix.spinnaker.proto.stats.Event.newBuilder()
+              .setSpinnakerInstance(spinnakerInstance);
+
+      telemetryService.sendMessage(JsonFormat.printer().print(eventProto));
```

Comment:
What's the failure mode if the back-end of the telemetry service is down or becomes slow?

If it's down and errors quickly, I suppose the worst that will happen is that we'll catch and log a lot of exceptions (which could still lead to a lot of log noise).

But what about the case where the service is slow?  (Or more likely, connections time out because people are running in environments that don't have access to the public internet.)

I don't see any timeout or circuit-breaking here.  Will this just eat up all of echo's connection pool in the event that connections time out?

Reply:
Added some connection and read timeout settings, though I think it would be nice to "autodetect" an airgapped env (say with an HTTP HEAD call to the endpoint?) and automatically disable itself if outbound traffic is disabled - but maybe I'm over thinking it and it's not that big of a deal and admins who are annoyed by it can opt out.